File: app/src/main/java/com/elementary/tasks/core/views/PrefsView.java
package com.elementary.tasks.core.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.drawable.Drawable;
import android.os.Build;
import androidx.annotation.DrawableRes;
import android.util.AttributeSet;
import android.view.View;
import android.widget.CheckBox;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.elementary.tasks.R;
import com.elementary.tasks.core.utils.LogUtil;
import com.elementary.tasks.core.utils.Module;

import java.util.ArrayList;
import java.util.List;

import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.SwitchCompat;

/**
 * Copyright 2016 Nazar Suhovich
 * <p/>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * <p/>
 * http://www.apache.org/licenses/LICENSE-2.0
 * <p/>
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
public class PrefsView extends RelativeLayout {

    private static final int CHECK = 0;
    private static final int SWITCH = 1;
    private static final int VIEW = 2;
    private static final int TEXT = 3;

    private CheckBox checkBox;
    private SwitchCompat switchView;
    private TextView title;
    private TextView detail;
    private TextView prefsValue;
    private View dividerTop, dividerBottom, prefsView;

    private boolean isChecked;
    private boolean isForPro;
    private int viewType = CHECK;

    private List<PrefsView> mDependencyViews = new ArrayList<>();
    private List<PrefsView> mReverseDependencyViews = new ArrayList<>();
    private List<OnCheckedListener> mOnCheckedListeners = new ArrayList<>();

    public PrefsView(Context context) {
        super(context);
        init(context, null);
    }

    public PrefsView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs);
    }

    public PrefsView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(context, attrs);
    }

    private void init(Context context, AttributeSet attrs) {
        View.inflate(context, R.layout.view_prefs, this);
        setDescendantFocusability(FOCUS_BLOCK_DESCENDANTS);
        title = findViewById(R.id.prefsPrimaryText);
        detail = findViewById(R.id.prefsSecondaryText);
        prefsValue = findViewById(R.id.prefsValue);
        checkBox = findViewById(R.id.prefsCheck);
        dividerTop = findViewById(R.id.dividerTop);
        dividerBottom = findViewById(R.id.dividerBottom);
        prefsView = findViewById(R.id.prefsView);
        switchView = findViewById(R.id.prefs_switch);
        if (attrs != null) {
            TypedArray a = context.getTheme().obtainStyledAttributes(
                    attrs, R.styleable.PrefsView, 0, 0);
            String titleText = "";
            String detailText = "";
            String valueText = "";
            boolean divTop = false;
            boolean divBottom = false;
            int res = 0;
            try {
                titleText = a.getString(R.styleable.PrefsView_prefs_primary_text);
                detailText = a.getString(R.styleable.PrefsView_prefs_secondary_text);
                valueText = a.getString(R.styleable.PrefsView_prefs_value_text);
                divTop = a.getBoolean(R.styleable.PrefsView_prefs_divider_top, false);
                divBottom = a.getBoolean(R.styleable.PrefsView_prefs_divider_bottom, false);
                isForPro = a.getBoolean(R.styleable.PrefsView_prefs_pro, false);
                viewType = a.getInt(R.styleable.PrefsView_prefs_type, CHECK);
                res = a.getInt(R.styleable.PrefsView_prefs_view_resource, 0);
            } catch (Exception e) {
                LogUtil.e("PrefsView", "There was an error loading attributes.", e);
            } finally {
                a.recycle();
            }
            setTitleText(titleText);
            setDetailText(detailText);
            setDividerTop(divTop);
            setDividerBottom(divBottom);
            setView();
            setValueText(valueText);
            setViewResource(res);
        }
        setChecked(isChecked());
        setVisible();
    }

    public void setOnCheckedListener(OnCheckedListener listener) {
        this.mOnCheckedListeners.add(listener);
    }

    public void setForPro(boolean forPro) {
        isForPro = forPro;
        setVisible();
    }

    public void setDependentView(PrefsView view) {
        if (view != null) {
            mDependencyViews.add(view);
            view.setOnCheckedListener(checked -> checkDependency());
        }
        checkDependency();
    }

    private void checkDependency() {
        boolean enable = true;
        for (PrefsView prefsView : mDependencyViews) {
            if (!prefsView.isChecked()) {
                enable = false;
                break;
            }
        }
        setEnabled(enable);
    }

    public void setReverseDependentView(PrefsView view) {
        if (view != null) {
            mReverseDependencyViews.add(view);
            view.setOnCheckedListener(checked -> checkReverseDependency());
        }
        checkReverseDependency();
    }

    private void checkReverseDependency() {
        boolean enable = true;
        for (PrefsView prefsView : mReverseDependencyViews) {
            if (prefsView.isChecked()) {
                enable = false;
                break;
            }
        }
        setEnabled(enable);
    }

    private void setVisible() {
        if (isForPro) {
            if (Module.isPro()) {
                setVisibility(VISIBLE);
            } else {
                setVisibility(GONE);
            }
        } else {
            setVisibility(VISIBLE);
        }
    }

    private void setView() {
        hideAll();
        switch (viewType) {
            case CHECK:
                checkBox.setVisibility(VISIBLE);
                break;
            case SWITCH:
                switchView.setVisibility(VISIBLE);
                break;
            case TEXT:
                prefsValue.setVisibility(VISIBLE);
                break;
            case VIEW:
                prefsView.setVisibility(VISIBLE);
                break;
        }
    }

    private void hideAll() {
        checkBox.setVisibility(GONE);
        switchView.setVisibility(GONE);
        prefsValue.setVisibility(GONE);
        prefsView.setVisibility(GONE);
    }

    public void setTitleText(String text) {
        title.setText(text);
    }

    public void setDetailText(String text) {
        if (text == null) {
            detail.setVisibility(GONE);
            return;
        }
        detail.setText(text);
        detail.setVisibility(VISIBLE);
    }

    public void setValue(int value) {
        prefsValue.setText(String.valueOf(value));
    }

    public void setValueText(String text) {
        prefsValue.setText(text);
    }

    public void setViewResource(@DrawableRes int resource) {
        if (resource != 0) {
            Drawable drawableTop;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                drawableTop = getContext().getDrawable(resource);
            } else {
                drawableTop = AppCompatResources.getDrawable(getContext(), resource);
            }
            prefsView.setBackground(drawableTop);
        }
    }

    public void setChecked(boolean checked) {
        this.isChecked = checked;
        if (viewType == CHECK) checkBox.setChecked(checked);
        else if (viewType == SWITCH) switchView.setChecked(checked);
        if (mOnCheckedListeners != null) {
            for (OnCheckedListener listener : mOnCheckedListeners) {
                listener.onCheckedChange(checked);
            }
        }
    }

    @Override
    public void setEnabled(boolean enabled) {
        super.setEnabled(enabled);
        switchView.setEnabled(enabled);
        checkBox.setEnabled(enabled);
        prefsView.setEnabled(enabled);
        prefsValue.setEnabled(enabled);
        detail.setEnabled(enabled);
        title.setEnabled(enabled);
    }

    public boolean isChecked() {
        return isChecked;
    }

    public void setDividerTop(boolean divider) {
        if (divider) {
            dividerTop.setVisibility(VISIBLE);
        } else {
            dividerTop.setVisibility(GONE);
        }
    }

    public void setDividerBottom(boolean divider) {
        if (divider) {
            dividerBottom.setVisibility(VISIBLE);
        } else {
            dividerBottom.setVisibility(GONE);
        }
    }

    public interface OnCheckedListener {
        void onCheckedChange(boolean checked);
    }
}

File: app/src/main/java/com/elementary/tasks/core/views/PrefsView.kt
package com.elementary.tasks.core.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.text.TextUtils
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.RelativeLayout
import androidx.annotation.ColorInt
import androidx.annotation.DrawableRes
import com.elementary.tasks.BuildConfig
import com.elementary.tasks.R
import com.elementary.tasks.core.binding.views.PrefsViewBinding
import com.elementary.tasks.core.utils.Module
import com.elementary.tasks.core.utils.ui.gone
import com.elementary.tasks.core.utils.ui.transparent
import com.elementary.tasks.core.utils.ui.visible
import timber.log.Timber

class PrefsView : RelativeLayout {

  private lateinit var binding: PrefsViewBinding

  var isChecked: Boolean = false
    set(checked) {
      field = checked
      if (viewType == CHECK) {
        binding.prefsCheck.isChecked = checked
      } else if (viewType == SWITCH) {
        binding.prefsSwitch.isChecked = checked
      }
      refreshDetailText()
      for (listener in mOnCheckedListeners) {
        listener.onCheckedChange(checked)
      }
    }
  private var isForPro: Boolean = false
  private var isTest: Boolean = false
  private var mDependentValue: Boolean? = null
  private var viewType = CHECK
  private var mOnText: String? = null
  private var mOffText: String? = null
  private var mSecondaryText: String? = null
  private val mDependencyViews = ArrayList<PrefsView>()
  private val mReverseDependencyViews = ArrayList<PrefsView>()
  private val mOnCheckedListeners = ArrayList<OnCheckedListener>()

  private val isCheckable: Boolean
    get() = viewType == CHECK || viewType == SWITCH

  constructor(context: Context) : super(context) {
    init(context, null)
  }

  constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
    init(context, attrs)
  }

  constructor(context: Context, attrs: AttributeSet, defStyle: Int) : super(
    context,
    attrs,
    defStyle
  ) {
    init(context, attrs)
  }

  private fun init(context: Context, attrs: AttributeSet?) {
    View.inflate(context, R.layout.view_prefs, this)
    binding = PrefsViewBinding(this)
    descendantFocusability = ViewGroup.FOCUS_BLOCK_DESCENDANTS

    binding.progressView.gone()

    if (attrs != null) {
      val a = context.theme.obtainStyledAttributes(
        /* set = */ attrs,
        /* attrs = */ R.styleable.PrefsView,
        /* defStyleAttr = */ 0,
        /* defStyleRes = */ 0
      )
      var titleText = ""
      var valueText = ""
      var divTop = false
      var divBottom = false
      var res = 0
      var iconId = 0
      try {
        titleText = a.getString(R.styleable.PrefsView_prefs_primary_text) ?: ""
        mSecondaryText = a.getString(R.styleable.PrefsView_prefs_secondary_text)
        mOnText = a.getString(R.styleable.PrefsView_prefs_secondary_text_on)
        mOffText = a.getString(R.styleable.PrefsView_prefs_secondary_text_off)
        valueText = a.getString(R.styleable.PrefsView_prefs_value_text) ?: ""
        divTop = a.getBoolean(R.styleable.PrefsView_prefs_divider_top, false)
        divBottom = a.getBoolean(R.styleable.PrefsView_prefs_divider_bottom, false)
        isForPro = a.getBoolean(R.styleable.PrefsView_prefs_pro, false)
        isTest = a.getBoolean(R.styleable.PrefsView_prefs_isTest, false)
        viewType = a.getInt(R.styleable.PrefsView_prefs_type, CHECK)
        res = a.getInt(R.styleable.PrefsView_prefs_view_resource, 0)
        iconId = a.getResourceId(R.styleable.PrefsView_prefs_icon, 0)
        val primaryColor = a.getColor(R.styleable.PrefsView_prefs_primary_text_color, -1)
        if (primaryColor != -1) {
          binding.prefsPrimaryText.setTextColor(primaryColor)
        }
      } catch (e: Exception) {
        Timber.d("init: ${e.message}")
      } finally {
        a.recycle()
      }
      if (iconId != 0 && SHOW_ICON) {
        binding.iconView.visible()
        binding.iconView.setImageResource(iconId)
        binding.iconView.imageTintList = binding.prefsPrimaryText.textColors
      } else {
        binding.iconView.transparent()
      }
      setTitleText(titleText)
      setDividerTop(divTop)
      setDividerBottom(divBottom)
      setView()
      setValueText(valueText)
      setViewResource(res)
    }
    setDetailText(mSecondaryText)
    isChecked = isChecked
    setVisible()
  }

  fun setLoading(isLoading: Boolean) {
    if (isLoading) {
      binding.progressView.visible()
    } else {
      binding.progressView.gone()
    }
  }

  fun setCustomViewClickListener(onClickListener: OnClickListener) {
    binding.prefsView.setOnClickListener(onClickListener)
  }

  fun setOnCheckedListener(listener: OnCheckedListener) {
    this.mOnCheckedListeners.add(listener)
  }

  fun setForPro(forPro: Boolean) {
    isForPro = forPro
    setVisible()
  }

  fun setDependentView(view: PrefsView) {
    mDependencyViews.add(view)
    view.setOnCheckedListener(object : OnCheckedListener {
      override fun onCheckedChange(checked: Boolean) {
        checkDependency()
      }
    })
    checkDependency()
  }

  fun setDependentValue(b: Boolean) {
    mDependentValue = b
    checkDependency()
    checkReverseDependency()
    checkDependentValue()
  }

  private fun checkDependency() {
    var enable = true
    for (prefsView in mDependencyViews) {
      if (!prefsView.isChecked) {
        enable = false
        break
      }
    }
    isEnabled = enable
    checkDependentValue()
  }

  private fun checkDependentValue() {
    val dep = mDependentValue ?: return
    if (isEnabled) {
      isEnabled = dep
    }
  }

  fun setReverseDependentView(view: PrefsView) {
    mReverseDependencyViews.add(view)
    view.setOnCheckedListener(object : OnCheckedListener {
      override fun onCheckedChange(checked: Boolean) {
        checkReverseDependency()
      }
    })
    checkReverseDependency()
  }

  private fun checkReverseDependency() {
    var enable = true
    for (prefsView in mReverseDependencyViews) {
      if (prefsView.isChecked) {
        enable = false
        break
      }
    }
    isEnabled = enable
    checkDependentValue()
  }

  private fun setVisible() {
    visibility = if (isTest) {
      if (BuildConfig.DEBUG) {
        View.VISIBLE
      } else {
        View.GONE
      }
    } else if (isForPro) {
      if (Module.isPro) {
        View.VISIBLE
      } else {
        View.GONE
      }
    } else {
      View.VISIBLE
    }
  }

  private fun setView() {
    hideAll()
    when (viewType) {
      CHECK -> binding.prefsCheck.visible()
      SWITCH -> binding.prefsSwitch.visible()
      TEXT -> binding.prefsValue.visible()
      VIEW -> binding.prefsView.visible()
    }
  }

  private fun hideAll() {
    binding.prefsCheck.gone()
    binding.prefsSwitch.gone()
    binding.prefsValue.gone()
    binding.prefsView.gone()
  }

  private fun setTitleText(text: String) {
    binding.prefsPrimaryText.text = text
  }

  fun setDetailText(text: String?) {
    if (isCheckable && hasOnOff()) {
      if (isChecked) {
        binding.prefsSecondaryText.text = mOnText
      } else {
        binding.prefsSecondaryText.text = mOffText
      }
      binding.prefsSecondaryText.visible()
    } else {
      if (TextUtils.isEmpty(text)) {
        binding.prefsSecondaryText.gone()
      } else {
        binding.prefsSecondaryText.text = text
        binding.prefsSecondaryText.visible()
      }
    }
  }

  fun setValue(value: Int) {
    binding.prefsValue.text = value.toString()
  }

  fun setValueText(text: String) {
    binding.prefsValue.text = text
  }

  fun setViewColor(@ColorInt color: Int) {
    if (color != 0) {
      binding.prefsView.setBackgroundColor(color)
    }
  }

  fun setViewResource(@DrawableRes resource: Int) {
    if (resource != 0) {
      binding.prefsView.setImageResource(resource)
    }
  }

  fun setViewTintColor(@ColorInt color: Int) {
    if (color != 0) {
      binding.prefsView.setColorFilter(color)
    }
  }

  fun setViewDrawable(drawable: Drawable?) {
    binding.prefsView.setImageDrawable(drawable)
  }

  fun setViewBitmap(bitmap: Bitmap?) {
    binding.prefsView.setImageBitmap(bitmap)
  }

  private fun refreshDetailText() {
    if (isCheckable && hasOnOff()) {
      setDetailText(mSecondaryText)
    }
  }

  override fun setEnabled(enabled: Boolean) {
    super.setEnabled(enabled)
    binding.prefsSwitch.isEnabled = enabled
    binding.prefsCheck.isEnabled = enabled
    binding.prefsView.isEnabled = enabled
    binding.prefsValue.isEnabled = enabled
    binding.prefsSecondaryText.isEnabled = enabled
    binding.prefsPrimaryText.isEnabled = enabled
    binding.iconView.isEnabled = enabled
  }

  private fun setDividerTop(divider: Boolean) {
    if (divider) {
      binding.dividerTop.visible()
    } else {
      binding.dividerTop.gone()
    }
  }

  private fun setDividerBottom(divider: Boolean) {
    if (divider) {
      binding.dividerBottom.visible()
    } else {
      binding.dividerBottom.gone()
    }
  }

  private fun hasOnOff(): Boolean {
    return mOffText != null && mOnText != null
  }

  interface OnCheckedListener {
    fun onCheckedChange(checked: Boolean)
  }

  companion object {

    private const val CHECK = 0
    private const val SWITCH = 1
    private const val VIEW = 2
    private const val TEXT = 3
    private const val SHOW_ICON: Boolean = true
  }
}
